File: backend/services/userService.js
/*
*Ella Muro
*24 April 2026
*Query Handler for Users MySQL Table
*/

//Importing the database connection
const pool = require('../db');
const bcrypt = require('bcrypt');

//Loading users from the MySQL table by their email
async function getUsersByEmail(email) {
    const[rows] = await pool.query(
        "SELECT * FROM users WHERE email = ?",
        [email]
    );

    //Returning users
    return rows[0];
}


//Method to create a new user
async function createUser(userName, email, password, docRole, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2) {
  
    const saltRounds = 10;

    const hashedPassword = await bcrypt.hash(password, saltRounds);

    await pool.query(
        "INSERT INTO users (userName, email, docRole, passwordHash, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
        [userName, email, docRole, hashedPassword, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2]
    );
}

//Method for loading users from the database
async function loadUsers() {
    const [rows] = await pool.query(
        "SELECT id, userName, email, docRole, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2 FROM users"
    )
    return rows;
}

//Method for updating a user record in the database
async function updateUser(id, userName, email, docRole, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2) {
    
    console.log("ID:", id);

    await pool.query(
        "UPDATE users SET "
                + "userName=?, "
                + "email=?, "
                + "docRole=?, "
                + "weeksAllowed=?, "
                + "prepicksAllowed=?, "
                + "priorityNumber=?, "
                + "prepicksPriorityNumber=?, "
                + "label=?, "
                + "displayName=?, "
                + "phoneNumber=?, "
                + "label2=? "
                + "WHERE id=?;",
        [userName, email, docRole, weeksAllowed, prepicksAllowed, priorityNumber, prepicksPriorityNumber, label, displayName, phoneNumber, label2, id]       
    )
}

//Method for deleting a user
async function deleteUser(id) {
    console.log("DELETE ID:", id);

    await pool.query(
        "DELETE FROM users WHERE id=?;",
        [id]
    )
}

//Method for changing password
async function changePassword(userId, newPassword, confirmNewPassword) {

    if(newPassword !== confirmNewPassword) {
        throw new Error("Passwords do not match");
    }

    const newHash = await bcrypt.hash(newPassword, 10);

    //Updating password in the database
    await pool.query(
        "UPDATE users SET passwordHash=? WHERE id=?",
        [newHash, userId]
    );
}

//Method to load a user by their id
async function getUserById(userId) {
    const[rows] = await pool.query(
        "SELECT * FROM users WHERE id=?",
        [userId]
    );

    //Returning users
    return rows[0];
}

module.exports = {getUsersByEmail, createUser, loadUsers, updateUser, deleteUser, changePassword, getUserById };

File: esavacation/lib/userService.dart
/*
*Ella Muro
*29 April 2026
*Class to Handle Users
*/

//Imports section
import 'user.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'role.dart';
import 'connectivityService.dart';

class UserService {
  final String baseUrl = "https://vacation-xhxd.onrender.com";

    //Method for loading users
    Future<List<User>> getUsers() async {

      try {
        final res = await http.get(Uri.parse("$baseUrl/users"));

        print("STATUS: ${res.statusCode}");
        print("BODY: ${res.body}");

        if(res.statusCode == 200) {
          final List data = jsonDecode(res.body);

          print(res.body);

          return data.map((e) => User.fromJson(e)).toList();
        }
        return [];
    } catch (error) {
      print("GET USERS ERROR: $error");
      ConnectivityService.instance.connectionFailed();
      throw Exception("No internet connection");
    }
  }

  //Method for creating users
  Future<bool> createUser({
    required String userName,
    required String email,
    required String password,
    required String docRole,
    required int weeksAllowed,
    required int prepicksAllowed,
    int? priorityNumber,
    int? prepicksPriorityNumber,
    String? label,
    required String displayName,
    required String phoneNumber,
    String? label2
  }) async {

    try {

    print("SENDING CREATE USER REQUEST...");

    final res = await http.post(
      Uri.parse("$baseUrl/users/register"),

      headers: {
        "Content-Type": "application/json",
      },

      body: jsonEncode({
        "userName": userName,
        "email": email,
        "password": password,
        "docRole": docRole,
        "weeksAllowed": weeksAllowed,
        "prepicksAllowed": prepicksAllowed,
        "priorityNumber": priorityNumber,
        "prepicksPriorityNumber": prepicksPriorityNumber,
        "label": label,
        "displayName": displayName,
        "phoneNumber": phoneNumber,
        "label2": label2
      }),
    ).timeout(Duration(seconds: 15));

      print("STATUS: ${res.statusCode}");
      print("BODY: ${res.body}");

      return res.statusCode == 200;
    } catch (error) {
      print("CREATE USER ERROR: $error");
      ConnectivityService.instance.connectionFailed();
      throw Exception("No internet connection");
    }
  }

  //Method for updating users
  Future<bool> updateUser({
    required int id,
    required String userName,
    required String email,
    required Role docRole,
    required int weeksAllowed, 
    required int prepicksAllowed,
    int? priorityNumber,
    int? prepicksPriorityNumber,
    String? label,
    required String displayName,
    required String phoneNumber,
    String? label2
  }) async {
    print ("UPDATING USER...");

    try {

    final res = await http.put(
      Uri.parse("$baseUrl/users/update"),

      headers: {
        "Content-Type": "application/json",
      },

      body: jsonEncode({
        "id": id,
        "userName": userName,
        "email": email,
        "docRole": docRole.name,
        "weeksAllowed": weeksAllowed,
        "prepicksAllowed": prepicksAllowed,
        "priorityNumber": priorityNumber,
        "prepicksPriorityNumber": prepicksPriorityNumber,
        "label": label,
        "displayName": displayName,
        "phoneNumber": phoneNumber,
        "label2": label2
      }),
    ).timeout(Duration(seconds: 15));

      print("STATUS: ${res.statusCode}");
      print("BODY: ${res.body}");

      return res.statusCode == 200;
    } catch (error) {
      print("UPDATE USER ERROR: $error");
      ConnectivityService.instance.connectionFailed();
      throw Exception("No internet connection");
    }
  }

  //Method for deleting a user
  Future<bool> deleteUserRecord({
    required int id,
  }) async {

    try {
        print("DELETING USER...");

        final res = await http.delete(
          Uri.parse("$baseUrl/users/$id"),
        ).timeout(Duration(seconds: 15));

      print("STATUS: ${res.statusCode}");
      print("BODY: ${res.body}");

      return res.statusCode == 200;

    } catch (error) {
      print("DELETE USER ERROR: $error");
      ConnectivityService.instance.connectionFailed();
      throw Exception("No internet connection");
    }
  }

  //Method for changing password
  Future<String> changePassword ({
    required int userId,
    required String currentPassword,
    required String newPassword,
  }) async {
    print("UPDATING PASSWORD..");

    try {
      final response = await http.put(
        Uri.parse("$baseUrl/users/updatePassword"),

        headers: {
          "Content-Type": "application/json",
        },

        body: jsonEncode({
          "userId": userId,
          "currentPassword": currentPassword,
          "newPassword": newPassword,
        }),
      ).timeout(Duration(seconds: 15));

      if(response.statusCode == 200) {
        return "Success";
      }

      final data = jsonDecode(response.body);
      return data["error"] ?? "Failed to change password";

    } catch (error) {
      ConnectivityService.instance.connectionFailed();
      return "Failed to connect to server";
    }
  }
}